ampd/src/stellar/http_client.rs

```diff
@@ -15,12 +17,41 @@ pub enum Error {
     Client,
     #[error("invalid tx hash")]
     TxHash,
+    #[error("failed to deserialize tx id")]
+    FailedTxIdDeserialization,
+}
+
+#[derive(Debug, Clone, Eq, Hash, Deserialize, Serialize, PartialEq)]
+pub struct StellarTxId(#[serde(deserialize_with = "deserialize_stellar_tx_id")] String);
+
+/// Deserializes a Stellar tx id from a 0x-prefixed tx id.
+fn deserialize_stellar_tx_id<'de, D>(deserializer: D) -> Result<String, D::Error>
```

Comment:
I think I understand the issue. The voting verifier emitted the tx id with a 0x prefix, but stellar doesn't like that. Is that right? If so, I don't think this is the best way to go about this, I find this approach rather confusing. I think a simpler way would be to just strip off the 0x prefix inside of transaction_response and transaction_responses. Or just change those functions to accept a type that strips the prefix off when converting from string. I find this automatic stripping as part of deserialization confusing.

Reply:
sorry for lack of info. Was discussing with Haiyi directly. Another [PR](https://github.com/axelarnetwork/axelar-amplifier/pull/647) was merged to unblock the testing. But this PR is meant to be an alternate fix, but it was failing to build due to a async fn requirement on the Deserialize error. Let me know if you know how to solve it

```
error[E0277]: `?` couldn't convert the error to `<__E as broadcaster::_::_serde::Deserializer<'de>>::Error`
  --> ampd/src/stellar/http_client.rs:24:44
   |
24 | #[derive(Debug, Clone, Eq, Hash, Deserialize, Serialize, PartialEq)]
   |                                            ^ the trait `From<error_stack::Report<<__E as broadcaster::_::_serde::Deserializer<'_>>::Error>>` is not implemented for `<__E as broadcaster::_::_serde::Deserializer<'de>>::Error`, which is required by `std::result::Result<StellarTxId, <__E as broadcaster::_::_serde::Deserializer<'de>>::Error>: FromResidual<std::result::Result<Infallible, error_stack::Report<<__E as broadcaster::_::_serde::Deserializer<'_>>::Error>>>`
   |
   = note: the question mark operation (`?`) implicitly performs a conversion on the error value using the `From` trait
   = note: required for `std::result::Result<StellarTxId, <__E as broadcaster::_::_serde::Deserializer<'de>>::Error>` to implement `FromResidual<std::result::Result<Infallible, error_stack::Report<<__E as broadcaster::_::_serde::Deserializer<'_>>::Error>>>`
   = note: this error originates in the derive macro `Deserialize` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: `?` couldn't convert the error to `<__D as broadcaster::_::_serde::Deserializer<'de>>::Error`
  --> ampd/src/stellar/http_client.rs:24:44
   |
24 | #[derive(Debug, Clone, Eq, Hash, Deserialize, Serialize, PartialEq)]
   |                                            ^ the trait `From<error_stack::Report<<__D as broadcaster::_::_serde::Deserializer<'_>>::Error>>` is not implemented for `<__D as broadcaster::_::_serde::Deserializer<'de>>::Error`, which is required by `std::result::Result<<http_client::_::<impl Deserialize<'de> for StellarTxId>::deserialize::__Visitor<'de> as broadcaster::_::_serde::de::Visitor<'de>>::visit_seq::__DeserializeWith<'_>, <__D as broadcaster::_::_serde::Deserializer<'de>>::Error>: FromResidual<std::result::Result<Infallible, error_stack::Report<<__D as broadcaster::_::_serde::Deserializer<'_>>::Error>>>`
   |
   = note: the question mark operation (`?`) implicitly performs a conversion on the error value using the `From` trait
   = note: required for `Result<__DeserializeWith<'_>, <__D as Deserializer<'de>>::Error>` to implement `FromResidual<std::result::Result<Infallible, error_stack::Report<<__D as broadcaster::_::_serde::Deserializer<'_>>::Error>>>`
   = note: the full name for the type has been written to '/Users/milap/work/axelar-amplifier/target/debug/deps/ampd-84bf7f1394b45ebf.long-type-9934518599439574020.txt'
   = note: consider using `--verbose` to print the full type name to the console
   = note: this error originates in the derive macro `Deserialize` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: `<D as broadcaster::_::_serde::Deserializer<'de>>::Error` cannot be sent between threads safely
  --> ampd/src/stellar/http_client.rs:32:58
   |
32 |     let tx_id: String = String::deserialize(deserializer)?;
   |                                                          ^ `<D as broadcaster::_::_serde::Deserializer<'de>>::Error` cannot be sent between threads safely
   |
   = help: the trait `std::marker::Send` is not implemented for `<D as broadcaster::_::_serde::Deserializer<'de>>::Error`, which is required by `std::result::Result<std::string::String, error_stack::Report<<D as broadcaster::_::_serde::Deserializer<'de>>::Error>>: FromResidual<std::result::Result<Infallible, <D as broadcaster::_::_serde::Deserializer<'_>>::Error>>`
   = note: required for `<D as broadcaster::_::_serde::Deserializer<'de>>::Error` to implement `error_stack::Context`
   = note: required for `error_stack::Report<<D as broadcaster::_::_serde::Deserializer<'de>>::Error>` to implement `From<<D as broadcaster::_::_serde::Deserializer<'_>>::Error>`
   = note: required for `std::result::Result<std::string::String, error_stack::Report<<D as broadcaster::_::_serde::Deserializer<'de>>::Error>>` to implement `FromResidual<std::result::Result<Infallible, <D as broadcaster::_::_serde::Deserializer<'_>>::Error>>`
help: consider further restricting the associated type
   |
30 |     D: Deserializer<'de>, <D as broadcaster::_::_serde::Deserializer<'de>>::Error: std::marker::Send
   |                         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

error[E0277]: `<D as broadcaster::_::_serde::Deserializer<'de>>::Error` cannot be shared between threads safely
```

Reply:
And to your larger point about the design, I think we need to fix the poll started event to not split tx id and event index. We should keep it as message id, and the appropriate handler should convert to the MessageIdFormat itself